packages/lib-classifier/src/plugins/drawingTools/models/marks/Rectangle/Rectangle.js

```diff
@@ -0,0 +1,99 @@
+import { addDisposer, getRoot, getParentOfType, isValidReference, types } from 'mobx-state-tree'
+import { autorun } from 'mobx'
+import { Rectangle as RectangleComponent } from '../../../components/'
+import { RectangleTool } from '@plugins/drawingTools/models/tools'
+
+import Mark from '../Mark'
+
+const MINIMUM_SIZE = 5
+
+const RectangleModel = types
+  .model('RectangleModel', {
+    x_center: types.maybe(types.number),
+    y_center: types.maybe(types.number),
+    width: types.maybe(types.number),
+    height: types.maybe(types.number)
+  })
+  .views(self => ({
+    get coords () {
+      return {
+        x: self.x_center,
+        y: self.y_center
+      }
+    },
+
+    deleteButtonPosition (scale) {
+      const BUFFER = 16
+      const x = self.x_center + self.width / 2 + (BUFFER / scale)
+      const y = self.y_center - self.height / 2
+      // TODO: check for out of bounds coordinates
+      return { x, y }
+    },
+
+    get isValid () {
+      return (self.width - MINIMUM_SIZE > 0)
+        && (self.height - MINIMUM_SIZE > 0)
+    },
+
+    get tool () {
+      return getParentOfType(self, RectangleTool)
+    },
+
+    get toolComponent () {
+      return RectangleComponent
+    }
```

Comment:
YIKES! I tried adding `get x () { return self.x_center }` (and variants thereof), but it causes a lot of wonk to happen, since the translation goes on top of the existing x_left, x_right, etc calculations.

I'm going to give this another shot today, but I think the `get x` and `get y` might best be done in a separate PR dedicated to adding rotations to the Rectangle, since that PR will likely need to update a bunch of the SVG calculations in any case.

Reply:
Definitely a future enhancement. It'll be needed when rotation is added, but that changes all the drag interactions anyway (since `x,y` in the rectangle's frame of reference no longer line up with `x,y` for the SVG.)

Reply:
Jinx!

Reply:
Yeah, I think I need to save the `get x` and `get y` for a different PR, because it also requires changing the Delete button.

So in the Rectangle Tool's calculations, I can do this...

```
// Actual data values
const x_left = x_center - width / 2
const x_right = x_center + width / 2
const y_top = y_center - height / 2
const y_bottom = y_center + height / 2

// SVG values, which factor in that the Rectangle's parent <g> will have
// transform="translate(x, y)"
const svg_left = -width / 2
const svg_right = width / 2
const svg_top = -height / 2
const svg_bottom = height / 2

const _onFinish = onFinish || function () { return true }

return (
  <g onPointerUp={active ? _onFinish : undefined}>
    <rect x={svg_left} y={svg_top} width={width} height={height} />
    <rect x={svg_left} y={svg_top} width={width} height={height} strokeWidth={GRAB_STROKE_WIDTH / scale} strokeOpacity='0' />
    {active &&
      <DragHandle
        scale={scale}
        x={svg_left}
        y={svg_top}
        dragMove={(e, d) => 
          onHandleDrag({
            x_left: x_left + d.x,
            x_right: x_right,
            y_top: y_top + d.y,
            y_bottom: y_bottom,
          })
        }
      />}
    ... etc ...
  </g>
)
```

...and this is a basic way of separating what the Rectangle LOOKS like (visual presentation, with the SVG accounting for transform="translate") and what the TRUE DATA values are (i.e. x, y, width, height relative to the Subject)

![Screen Shot 2020-01-22 at 18 08 21](https://user-images.githubusercontent.com/13952701/72920931-32f0d000-3d42-11ea-80cd-6e1bb483db95.png)

Problem is, the Delete button already has its own `transform` shenanigans going on, and that has to be changed for this implementation of Rectangle calculations to work. 

_ Picture: the Delete (X) button being offset far too much, due to taking in its transform="translate" PLUS the Rectangle's transform="translate", the latter which only exists because we've implemented 'get x' and 'get y'_

This can be handled, but best done in a separate PR, since it'll affect way more marks than the Rectangle.

Reply:
The Ellipse PR #1360, which does have rotation, has examples of this. The centre of your shape moves to `( 0, 0 )` and the delete button position is calculated relative to that. Rectangles might move the top left corner to `( 0, 0 )` on translate. SVG doesn't seem to have any standards as to how translations are applied.

I do remember that the Ellipse PR changes the behaviour of `dragMove` because you can't handle rotation with just `dx, dy`. With that PR, I think it becomes `dragMove({ x, y }, { x: dx, y: dx })`. It's been a while since I wrote it.